src/libostree/ostree-fetcher-soup.c

```diff
@@ -386,8 +385,6 @@ start_pending_request (ThreadClosure *thread_closure,
   OstreeFetcherPendingURI *pending;
   GCancellable *cancellable;
 
-  g_assert_cmpint (g_hash_table_size (thread_closure->outstanding), <, thread_closure->max_outstanding);
```

Comment:
While I'm generally happy to remove an `assert` that the code is not doing anything to prevent, note that the curl fetcher [has a similarly hardcoded assertion](https://github.com/ostreedev/ostree/blob/master/src/libostree/ostree-fetcher-curl.c#L870).

I'm also curious why the fetcher is getting to 24 requests when the pull queue code is supposed to stop at 8. Oh, I think I might see it. Most fetches check `fetcher_queue_is_full` before calling `start_fetch` or `start_fetch_deltapart`.

However, all the initial requests are started with `initiate_request`. In that function, if no delta is found then you go into `queue_scan_one_metadata_object`, which eventually checks `fetcher_queue_is_full`. But if a delta is found, then you go to `initiate_delta_request`, which calls `_ostree_fetcher_request_to_membuf` without checking that the fetcher queue is full.

So, I think you can hit this problem if you pull more than 24 deltas. The solution would appear to be that a queue is needed for superblocks.

Reply:
Thanks, I updated it to drop the assertion from libcurl too.  You're right about the code chain involved here...I think it's fixable, we'd need to maintain integer offsets as iterators over the bits that call `initiate_request()`, but doing too many HTTP requests is better than crashing for now.

Reply:
Sure. I think you could refactor a bit with the existing scheme using a `pending_fetch_superblock` set and `start_superblock_fetch` utility function that does the actual `_ostree_fetcher_request_to_membuf` call. Then you'd just check `fetcher_queue_is_full` in `initiate_delta_request` and either add to the set or call `start_superblock_fetch`.